Fix Dedup Issues, Fix Camper Upload Issues, Enable Chugim CSV Upload, Edit Edah Term, Bunk Sort

JonahZW opened the pull request:
## Work Done
### Dedup
* Fixed issues where some chugim didn't deduplicate
* Only display top half of dedup matrix -- if chug A cannot be given to someone with chug B, implies that chug B also cannot be given to someone with chug A
* If Chug A is deduped with Chug B via the edit page for Chug A, the dedup is shown on the edit page for Chug B as well
* Significantly sped up saving time for dedup matrix page

### Camper Upload Fix
* Much better error validation provided when data cannot be saved/loaded
* Template CSV available for download
* Better formatting to describe requirements

### Chug CSV Upload
* Similar mechanism to uploading campers implemented -- new chugim can be bulk added via CSV
* Template available for download
    * Chug, min/max, department, description, etc
    * columns for each block and edah where users just need to mark within the corresponding cell to indicate which block(s)/edah(ot) the chug is available for
* Adds chugim to edot_for_chug and chug_instances table as they are created

### Other Improvements
* Ability to edit the edah term (Wisconsin spells it "eidah" and "eidot")
* Bunks are now sorted numerically, not alphabetically (e.g. 1-2-3-4-10-11 instead of 1-10-11-2-3-4)

## Required SQL
```
-- add edah term, chug_importer to admin_data
alter table admin_data add edah_term_singular varchar(50) default 'edah' not null after block_term_plural;
alter table admin_data add edah_term_plural varchar(50) default 'edot' not null after edah_term_singular;
alter table admin_data add enable_chugim_importer tinyint(1) default 0 not null after edah_term_plural;

-- remove duplicates from chug_dedup_instances_v2 and add a unique constraint
-- note: need to add then remove an id column to make this easier
ALTER TABLE chug_dedup_instances_v2 ADD id INT PRIMARY KEY AUTO_INCREMENT;
DELETE c1 FROM chug_dedup_instances_v2 c1 INNER JOIN chug_dedup_instances_v2 c2 WHERE c1.id > c2.id AND c1.left_chug_id = c2.left_chug_id and c1.right_chug_id = c2.right_chug_id;
ALTER TABLE chug_dedup_instances_v2 DROP COLUMN id;
alter table chug_dedup_instances_v2 ADD CONSTRAINT UNIQUE `unique` (`left_chug_id`, `right_chug_id`);

alter table chug_groups drop foreign key chug_groups_ibfk_1;
alter table chug_groups add constraint `chug_groups_ibfk_1` FOREIGN KEY (`active_block_id`) REFERENCES `blocks` (`block_id`) ON DELETE SET NULL ON UPDATE CASCADE;
```

JonahZW committed:
Fix Dedup Issues, Optimize

 * Fixes issue where some chugim wouldn't allow duplication, even when designated
 * Only displays upper half of dedup matrix (assumes all relationships go both ways)
 * When a chug is set to de-duped with another, the edit-chug page for both chugim automatically displays appropriately
 * Significantly speeds up process of saving updates to dedup matrix (one large SQL query as opposed to individual ones for each pair)
* Adds functionality to dbConn.php to allow more flexibility - better control for WHERE clauses (default is same) and also allows multiple inserts to be compiled with the same dbConn request

JonahZW committed:
Update Bunk Sort Order

Sort bunks by number, not alphabetically

JonahZW committed:
Allow editing edah term

Ramah Wisconsin spells "edah" as "eidah;" created new admin option to modify spelling of this term.

SQL to update DB:

alter table admin_data add edah_term_singular varchar(50) default 'edah' not null after block_term_plural;
alter table admin_data add edah_term_plural varchar(50) default 'edot' not null after edah_term_singular;

JonahZW committed:
Edah Term DB Updates

Updates SQL files to support new editable edah term

JonahZW committed:
Fix Camper Upload; Add Chugim Upload

 * fixed issues with camper CSV upload (added error validation and more clarity to upload page)
 * created ability to upload chugim as a CSV

Required SQL to update db:
alter table admin_data add enable_chugim_importer tinyint(1) default 0 not null after edah_term_plural;

JonahZW committed:
Delete Comment

dlobron commented on chugbot/camperUpload.php:
```diff
@@ -4,26 +4,87 @@
+    } else if ($cparts[0] == "downloadSample") {
+        // orig adapted from https://stackoverflow.com/a/16251849
+        header('Content-Type: application/csv');
+        header('Content-Disposition: attachment; filename="camperSample.csv";');
```
This is a great idea! Thank you for adding.

JonahZW replied:
Thanks!

dlobron commented on chugbot/camperUpload.php:
```diff
@@ -4,26 +4,87 @@
+        $count = 1;
+        while ($row = mysqli_fetch_array($result, MYSQLI_ASSOC)) {
+            $firstPref = null;
+            if ($count % 4 == 0) {$firstPref = true;}
```
Will this break if the camp uses a different number of chugim in the choices? For example, Palmer usually uses 6 chugim. Also, it seems like this will leave firstPref set to true as soon as count hits 4- is that intentional?

JonahZW replied:
I don't think it should break - this section is generating fake data for the sample camper CSV to demonstrate what it could/should look like.

`$count` tracks the number of rows of sample data which have been created. `$firstPref` is a boolean which is used to indicate (in the sample data) if a camper should always get their first choice ("needs_first_choice"); every four "campers" in the sample data will have this flag set (hence the `$count` modulo 4).

Does this make more sense? Do you have ideas for how I could/should better document it?

dlobron replied:
Oh, I see- thank you for clarifying. I missed that $firstPref = null. This should be fine!

JonahZW replied:
Added a comment clarifying it!

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+    exit();
+}
+
+// map edah name, block name, chug group name to relevant ids
```
In a future PR, we should move this functionality to a utility, since it's done in a few places. Definitely not needed in this PR.

JonahZW replied:
Do you think it would be best to make a function which returns a boolean for the term?

e.g. this chunk would become something like:
```
$enableChugimImporter = check_enabled("enable_chugim_importer");
if (!$enableChugimImporter) {
    $redirUrl = urlIfy("../staffHome.php?from=chugimUpload.php");
    header("Location: $redirUrl");
    exit();
}
```

dlobron replied:
Yes, exactly- I was thinking we could hoist this to a function in utils.php. We can do that in a future PR, though, if you prefer.

JonahZW replied:
Done! [9391a3a](https://github.com/dlobron/ChugBot/pull/74/commits/9391a3ad5d2ea322bc3cb8dbecdf06a1a7abd9e9)

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+            $blockEdahLen = sizeof($block_name_to_id) + sizeof($edah_name_to_id) - 1;
+            $blockEdah = str_repeat(",", $blockEdahLen);
+            $temp = $blockEdahLen - random_int(0,3);
+            while ($temp >= 0) {
```
This while loop confused me slightly- can you add a comment that explains what it does? I just want to make sure there's no risk of an infinite loop.

JonahZW replied:
Ah, yeah, I see - the sample chug list to upload has columns for each block and edah; marking a value in the column sets it as available for that chug (e.g. if the row for chug "Ropes" has a "1" in the column for every edah, campers in every edah will be able to signup for "ropes." If it only has marks in the columns for Magshimim-Nivonim, only the b-side campers would be able to sign up for it).

This while loop creates fake entries for those columns. So it works as a CSV, a string is built with the number of commas as there are blocks/edot (e.g. ",,,,,,,,,,"). Then, `$temp` is set to an index near the end of the string, and inserts in a 1 (e.g. ",,,,,,,,,1,"). `$temp` is decremented by a small, random amount to add in more values (e.g. making the string be ",,,,,,,1,,1," -> ",,,,1,,,1,,1," -> ",,1,,1,,,1,,1,"). Going from the end of the string means we know subtracting anything from `$temp` will either be between 2 commas or beyond the beginning of the string (which will break the loop because `$temp` will be less than 0).

I'll definitely add a comment explaining this; does this explanation make sense and help?

dlobron replied:
Thank you for the explanation. I think I understand it, but a comment would be great. The mechanics of it make sense, but I just want to make sure I understand the purpose of building the string with the random 1's.

JonahZW replied:
Here's a screenshot of the resulting CSV (opened in Numbers):
<img width="1281" alt="image" src="https://github.com/user-attachments/assets/76228292-739c-4ba0-8db6-bad3fffb3a0a" />

When going through the regular form to add a chug, options like name, min, max, description, chug group, etc all can only hold one value. Edot and blocks both can have multiple true options at once (e.g. we can have a chug for multiple edot during multiple blocks). The CSV file being uploaded is structured to have one column per each of those potential options (each edah has a column and each block has one). If there is a "1" in the cell for a certain column/row, the chug is available for that edah and/or block (e.g. Chug 0 is available in Sessions A, B, and C for Garinim). The benefit of having the random locations is it demos how the uploaded file can look and provides a starting point for staff preparing to upload chugim.

JonahZW replied:
I also added a comment within the loop indicating this is the purpose of that code section

dlobron replied:
Thank you! Sounds good.

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+        // min is non-negative number
+        $min = (int)$chug['min'];
+        if ($chug['min'] != null & $chug['min'] != $min) {
+            $error = "min ('" . $chug['min'] . "') is not a number";
```
nit: Maybe say "not an integer", since the cast could change a float to an int (unlikely, but we should handle it). Also, since we do this below for $max, consider hoisting to a function called, for example, convert_to_positive_int, which can return null if the conversion fails.

JonahZW replied:
Good idea, thanks!

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+        }
+    }
+
+    // determine next chug_id
```
Can you briefly explain why you need to access the next ID? It's generally best practice to let the database auto-increment, rather than inserting explicitly.

JonahZW replied:
When adding a new chug, we need to also add entries to the tables `chug_instances` (mapping chugim to blocks) and `edot_for_chug` (mapping chugim to edot). Those both rely on the chug_id. The camperUpload page was setup to make a queue of SQL commands which execute when we know there are no errors, and I mirrored that here.

The problem is that I don't know if we can get the chug_id from a chug which hasn't been pushed to the database so we can add to `chug_instances` and `edot_for_chug` as well.

Do you have ideas for other ways we would guarantee the rollback-ability if there are errors with the upload?

dlobron replied:
Ah, got it- hm. My general approach here was that the user would create the new chug and do the initial assignment of the chug to one or more blocks and edot, which could then be edited. 

I think I may need to wrap my head better around the purpose of this function. Can you say a bit more about how it works? Is the idea that we want to insert a chug "automatically", without asking the user to manually fill in the blocks and edot for the chug?

JonahZW replied:
Basically -  my overall goal with this page was to replicate the existing "Add Chug" page but make it possible to add chugim in bulk. I think it would be cumbersome to create the initial chugim and then need to go through the "Edit Chug" form for each one to set the edot and blocks; it would essentially remove the benefit of bulk uploading.

This way the user indicates via the CSV which edot/blocks the chug is available for; the chug is then fully created.

So yes, we want the user to be able to insert a batch of many chugim automatically so they don't need to later manually fill in blocks/edot

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+            continue;
+        }
+
+        // add chug to db
```
Would the dbConn->insertIntoTable method work here? It has a bind_param on line 200. If it doesn't work for you as-is, we could try changing it so it would work. That can also go in a future PR if you prefer, though.

JonahZW replied:
I modeled this heavily off of the camperUpload page where it was setup this way. I think the big benefit is it lets us make a "queue" of SQL commands which we later commit to the DB when we know there are no errors. I don't really want to have some chugim successfully upload and some fail, it seems better to me if it is "all or nothing."

In short, `dbConn->insertIntoTable` would do it immediately; this method seems to allow us to defer until we can confirm the time is right

dlobron replied:
Ah, got it- yup, that makes sense. In the future, we could perhaps make a dbConn that would take a queue of commands, and execute them in a batch that could be rolled back. This is great for now.

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,377 @@
+            }
+        }
+
+        // add edot_for_chug to db based on selected edot
```
Here and above/below, I'd recommend insertIntoTable if it works, or else we can fix that method in a future PR.

JonahZW replied:
See previous comment :)

dlobron commented on chugbot/dbConn.php:
```diff
@@ -73,28 +73,55 @@ public function addColVal($val, $type)
         array_push($this->colVals, $val);
+        // Determine if next column should be and/or for conditional - default AND
+        if ($andOr != "AND" && $andOr != "OR") {
+            $andOr != "AND";
```
Should this be `$andOr = "AND";`?

JonahZW replied:
YES! Thank you for catching this!

dlobron commented on chugbot/exclusionMatrix.html:
```diff
@@ -19,12 +19,13 @@
     href="https://cdn.datatables.net/u/dt/jqc-1.12.3,dt-1.10.12,fc-3.2.2,fh-3.1.2/datatables.min.css" />
-  <script type="text/javascript"
-    src="https://cdn.datatables.net/u/dt/jqc-1.12.3,dt-1.10.12,fc-3.2.2,fh-3.1.2/datatables.min.js"></script>
+<script src="https://cdn.datatables.net/2.1.8/js/dataTables.js"></script>
```
Much better- thank you for fixing this!

JonahZW replied:
Thanks!

dlobron reviewed (commented):
Amazing work, Jonah! Thank you so much for putting so much time and effort into this. I added a few comments and suggestions. Some of the suggestions can be in a future PR, and feel free to push back on any comments where I might have not understood something. Again, great work, and thank you again!

JonahZW committed:
Camper/Chug Upload Adjustments

Fixes based on comments

dlobron commented on chugbot/chugUpload.php:
```diff
@@ -0,0 +1,386 @@
+
+<p class="mb-1">Expand any of the below fields to view allowed values for each column</p>
+
+<div class="accordion mb-4" id="detailAccordion">
```
I like the accordion interface a lot. Would it be possible to write a function that would generate each <div> section, or are they different enough from one another that it's better to have them separate? I'm just trying to keep the lines of code as small as possible and be modular where we can.

dlobron replied:
(same for designSchedules.php below)

JonahZW replied:
That's a good idea, I'm going to think about this a bit.

I think we could - it would probably need the following parameters (at least):
* `title` (for the heading)
* `body` (contents of the accordion)
* `collapsed` (boolean saying if it's initially collapsed or open)
* `id` (number, title, something like that)

There may be more, but I'll try and do this tonight!

dlobron replied:
That sounds great! I'll mark this resolved in case you'd like to do that work in a separate PR - either way is OK!

JonahZW committed:
admin_data quick check

New utility function takes a parameter with a column name and responds with a boolean if there is a value in admin_data

dlobron commented:
LGTM!

JonahZW committed:
Bootstrap Accordion Utility

Implements a new class to easily create a Bootstrap 5.0 Accordion. Adopts current code to utilize this

JonahZW commented:
I think we should now be good - I just added a new class to functions.php which allows for easy creation of Bootstrap Accordions and changed all accordions to use this new class

JonahZW committed:
Separate chugim from chug groups in archive page

Also includes instructions to install mysql and mysqldump for the Docker testing environment

dlobron commented:
Great work, Jonah! Please go ahead and merge when needed. Thank you for making the functions!
